domain/application/state/application.go

```diff
@@ -1179,6 +1179,162 @@ WHERE a.name = $applicationName.name
+	var result []spaceAddress
+
+	app := applicationID{ID: appUUID}
+	stmt, err := st.Prepare(`
+WITH units_and_services AS (
+    SELECT
+        u.net_node_uuid AS net_node_uuid
+    FROM
+        unit u
+    WHERE
+        u.application_uuid = $applicationID.uuid
+    UNION ALL
+    SELECT
+        ks.net_node_uuid AS net_node_uuid
+    FROM
+        k8s_service ks
+    WHERE
+        ks.application_uuid = $applicationID.uuid
+)
+SELECT
+    ip.address_value AS &spaceAddress.address_value,
+    ip.type_id AS &spaceAddress.type_id,
+    ip.scope_id AS &spaceAddress.scope_id,
+    sn.space_uuid AS &spaceAddress.space_uuid
+FROM
+    units_and_services uas
+JOIN
+    link_layer_device lld ON lld.net_node_uuid = uas.net_node_uuid
+JOIN
+    ip_address ip ON ip.device_uuid = lld.uuid
+LEFT JOIN
+    subnet sn ON sn.uuid = ip.subnet_uuid;
+`, app, spaceAddress{})
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	err = tx.Query(ctx, stmt, app).GetAll(&result)
+	if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+		return nil, errors.Errorf("querying application %q space addresses: %w", appUUID, err)
+	}
+	return result, nil
+}
+
+func (st *State) hashAddressesAndEndpoints(addresses []spaceAddress, endpointBindings map[string]string) (string, error) {
+	if len(addresses) == 0 {
+		return "", nil
+	}
+
+	hash := sha256.New()
+	// Sort addresses by value, which is needed for the hash to be consistent.
+	sort.Slice(addresses, func(i, j int) bool {
+		return addresses[i].Value < addresses[j].Value
+	})
+	// Add the hash parts for each address.
+	for _, spaceAddress := range addresses {
+		if _, err := hash.Write([]byte(spaceAddress.Value)); err != nil {
+			return "", errors.Errorf("hashing address %q: %w", spaceAddress.Value, err)
+		}
+		addressType := ipaddress.UnMarshallAddressType(ipaddress.AddressType(spaceAddress.TypeID))
+		if _, err := hash.Write([]byte(addressType)); err != nil {
+			return "", errors.Errorf("hashing address type %q: %w", addressType, err)
+		}
+		addressScope := ipaddress.UnMarshallScope(ipaddress.Scope(spaceAddress.ScopeID))
+		if _, err := hash.Write([]byte(addressScope)); err != nil {
+			return "", errors.Errorf("hashing address scope %q: %w", addressScope, err)
+		}
+		spaceUUID := network.AlphaSpaceId
+		if spaceAddress.SpaceUUID.Valid {
+			spaceUUID = spaceAddress.SpaceUUID.String
+		}
+		if _, err := hash.Write([]byte(spaceUUID)); err != nil {
+			return "", errors.Errorf("hashing space uuid %q: %w", spaceUUID, err)
+		}
+	}
+	// Sort the endpoint bindings by key (endpoint name) and add each binding to
+	// the hash.
+	endpointNames := make([]string, 0, len(endpointBindings))
+	for name := range endpointBindings {
+		endpointNames = append(endpointNames, name)
+	}
+	sort.Strings(endpointNames)
+	for _, endpointName := range endpointNames {
+		if _, err := hash.Write(fmt.Appendf(nil, "%s:%s", endpointName, endpointBindings[endpointName])); err != nil {
+			return "", errors.Capture(err)
+		}
+	}
```

Comment:
Something about this doesn't sit right. I'd rather we marshalled into something then hashed that (string). We also, seem to be encoding from a int into a string to then write it out. We could just use the int as a byte.

Reply:
IMO it's more readable to have a hash constructed part by part like it's done here, the main reason being that both inputs have to be sorted and a pre-marshalled string would be confusing.
Regarding the IDs I agree with you, it's a good idea to just write the int IDs into the hash since we only care about a hash that changed with respect to a previous one, not its contents.